hazelcast/src/test/java/com/hazelcast/jet/impl/deployment/ProcessorClassLoaderTest.java

```diff
@@ -0,0 +1,223 @@
+        if (jarFile != null) {
+            jarFile.delete();
+            jarFile = null;
+        }
+        if (resourcesJarFile != null) {
+            resourcesJarFile.delete();
+            resourcesJarFile = null;
+        }
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        ResourceCollector.items().clear();
+
+        member = createHazelcastMember();
+        client = HazelcastClient.newHazelcastClient();
+        jet = client.getJet();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        if (client != null) {
+            client.shutdown();
+        }
+        if (member != null) {
+            member.shutdown();
+        }
+    }
+
+    private HazelcastInstance createHazelcastMember() throws MalformedURLException {
+        Config config = smallInstanceConfig();
+        config.getJetConfig().setResourceUploadEnabled(true);
+
+        // Create a member in a separate classloader without the test classes loaded
+        URL classesUrl = new File("target/classes/").toURI().toURL();
+        HazelcastAPIDelegatingClassloader classloader = new HazelcastAPIDelegatingClassloader(
+                new URL[]{classesUrl},
+                // Need to delegate to system classloader, which has maven dependencies like Jackson
+                ClassLoader.getSystemClassLoader()
+        );
+        HazelcastInstance member = HazelcastStarter.newHazelcastInstance(config, classloader);
+        return member;
+    }
+
+    @Test
+    public void testClassLoaderForBatchSource() throws Exception {
+        Pipeline p = Pipeline.create();
+        BatchSource<String> source = SourceWithClassLoader.batchSource(SOURCE_NAME);
+
+        p.readFrom(source).setLocalParallelism(1)
+         .writeTo(Sinks.list("test"));
+
+        JobConfig jobConfig = new JobConfig();
+        jobConfig.addCustomClasspath(source.name(), resourcesJarFile.getName());
+        jobConfig.addCustomClasspath(source.name(), jarFile.getName());
+        jet.newJob(p, jobConfig).join();
+
+        IList<Object> list = member.getList("test");
+        assertThat(list).contains("resource in jar");
+
+        assertThat(ResourceCollector.items()).containsExactly(
+                "Processor init resource in jar",
+                "Processor complete resource in jar"
+        );
+    }
+
+    @Test
+    public void testClassLoaderForStreamSource() throws Exception {
+        Pipeline p = Pipeline.create();
+        StreamSource<String> source = SourceWithClassLoader.streamSource(SOURCE_NAME);
+
+        p.readFrom(source)
+         .withoutTimestamps()
+         .setLocalParallelism(1)
+         .writeTo(Sinks.list("test"));
+
+        JobConfig jobConfig = new JobConfig();
+        jobConfig.addCustomClasspath(source.name(), resourcesJarFile.getName());
+        jobConfig.addCustomClasspath(source.name(), jarFile.getName());
+        jet.newJob(p, jobConfig);
+
+
+        Thread.sleep(5000);
```

Comment:
Maybe `assertJobStatusEventually(job, RUNNING)` instead?

Reply:
This sometimes hangs, because the processor actually completes, which completes the job (the state is COMPLETED). So I can use `job.join()` instead.